README: updating with schema information and updating instructions (#353)
unicode-org/conformance: README.md

# Unicode & CLDR Conformance Testing

This repository provides tools and procedures for verifying that an
implementation is working correctly according to the data-based
specifications. The tests are implemented on several platforms including NodeJS
(JavaScript), ICU4X (RUST), ICU4C, etc. Additional programming platforms may be
added to use the test driver framework.

One goal of this work is an easy-to-use framework for verifying that an
implementation of ICU functions agrees with the required behavior. When a test
passes in all implementations, it's a strong indication that results are
consistent across platforms.

A second goal is to find issues in particular implementations of ICU
functions. This is facilitated by generating data from CLDR (in most cases) or

> other trused sources. By highlighting differt results than expected,

Reviewer: ```suggestion
other trusted sources. By highlighting different results than expected,
```

implementation errors can be highlighted in ways that unit testing in each
platform cannot easily detect.

**Conformance Testing** is also known as **Data Driven Test (DDT)**. It focuses
on functions that accept data input such as numbers, date/time data, and other
basic information. The specifications indicate the expected output from
implementations when given the data and argument settings for each of the many
individual data items.

Note that these tests are only one part of the testing required for

> ICU-compliant libraries. Many additional tests are implemented in the

Reviewer: ```suggestion
ICU-compliant libraries. Many additional tests are implemented in the
unit tests and other tests of the library implementations themselves.
```

* Test verification uses ICU version information in the test output files for
  matching with the corresponding expected results. Verification output appears
  in the testResults subdirectory for each node, e.g. testOutput/rust/icu71.
  
* Schema checking uses explict descriptions of the data to be generated and used
in Conformance Testing. These include formats for test data, verification data,
and test output files.

### Terminology for Conformance Testing

> * **Platform: a library or softward that is executed to produce expected output

Reviewer: ```suggestion
* **Platform: a library or software that is executed to produce expected output
```

  such as formatted text. Examples include ICU4C, ICU4X, and NodeJS.

* **Component**: a type of test such as collation, list format, or plural

>   rules. Each platform may implement test for one or more compoenets.

Reviewer: ```suggestion
  rules. Each platform may implement tests for one or more components.
```

* **Version**: ICU and CLDR are released periodically with updated data and new

>   capabilities. ICU version are two digit numbers with optional fractilan part,

Reviewer: ```suggestion
  capabilities. ICU versions are two digit numbers with optional minor revision number per semantic versioning,
```

  e.g., ICU76 or CLDR76. A platform usually is implemented in multiple ICU

>   versions. Platforms should be updated to inclue3d each new ICU release.

Reviewer: ```suggestion
  versions. Platforms should be updated to include each new ICU release.
```

## Architectural Overview

Conceptually, there are four main functional units of the DDT implementation:

![Conceptual model of Data Driven Testing](./ddt_concept_model.png)

# How to use Conformance Testing on your computer

## Installing Comformance Testing

The Conformance projectis publicly available. It can be installed and executed
on most computer systems. Using GitHub, developers can investigate how the
executors run each of the components. And developers can make changes and create
pull requests to suggest updates to the projectd.

### Get the source from GitHub

1. Create a directory on the computer for installing and running
   Conformance. Note that this may require more that 4 Gbytes when all the tests
   are executed.

1. Next, get the code and data from the [Unicode Conformance GitHub
   site](https://github.com/unicode-org/conformance).

> 1. From the green Code button on Git but, select one of the options for cloning the

Reviewer: ```suggestion
1. From the green Code button on Git button, select one of the options for cloning the
```

project, or download the project as a ZIP file.

1. Unzip if needed. Then make sure that the directory contains directories for testdriver, testgen, schema, executors, verifier.

### Set up tools and execution environments

> Some setup is requirel to run Data Driven Testing code locally:

Reviewer: ```suggestion
Some setup is required to run Data Driven Testing code locally:
```

    the OS package that encapsulates the Python package
        ```
        sudo apt-get install python-jsonschema
        ```

Note: Setting up Python modules or other software may be better done using a
virtual environment such as using `pip env` for installation.

- Install the minimum version supported by ICU4X
    * The latest minimum supported supported Rust version ("MSRV") can be found in the
    [`rust-toolchain.toml` file](https://github.com/unicode-org/icu4x/blob/main/rust-toolchain.toml)
    * To view your current default Rust version (and other locally installed Rust versions):
        ```
        rustup show
        ```
    * To update to the latest Rust version:
        ```
        rustup update
        ```
  - Install `logrotate`
      ```
      sudo apt-get install logrotate
      ```

### Running End-To-End on a local computer

Running conformance is easy. Simply execute the script that first generates
data, runs executors, checks data against schema, and then runs the verifier,
creating the dashboard.

The standard script runs all components on all platforms with all ICU
version. Note that this takes a few minutes.

> `` bash generateDataAndRun.sh ``

Reviewer: Make this a code fence by doing a triple backquote on a separate line before and after the command

Output will be created in the directory TEMP_DATA.

To run a quicker version that uses only 100 test cases for each instance, run this:

``
bash genData100.sh

> ``

Reviewer: fix triple quotes in code fence here too

which does the same thing much faster, resulting in the directory TEMP_DATA_100.

### Viewing conformance results in a browser
HTML output is found in the subdirectory testReports. To visualize this on your
computer, start a webserver. For example:

``
python3 -m http.server 9000 &

> ``

Reviewer: ditto

Then open the file testResults/index.html under the testReports folder in either
TEMP_DATA or TEMP_DATA_100 or in a custom folder.

### How this works

Two main pieces are used to create conformance output:
* run_config.json which describes the platforms and components to be executed

* execution scripts that run all the steps of conformance testing. This file
  extracts configurations from run_config
  

>   * generateDataAndRun.sh runs tests across all platforms, components, and ICU

Reviewer: if this is not a sub-bullet, then unindent to be flush with the other points. if this is meant to indent, then indent by 4 spaces to make Markdown parsers happy

    versions.

*run_config.json* file is read by the execution scripts such as *generateDataAndRun.sh*

#### Execution script functions

The execution scripts perform several steps in sequence:

1. Special configuration to set up test environments for

    1. ICU4C versions

    1. NodeJS releases
    
    1. ICU4X using Rust

    1. Dart native and Dart web

1. Set the output directory as TEMP_DIR. Remove old data and create new output areas.

1. Generate all test data based on run_config's requirements

>    1. This includes checking each generated test set and expected results against schema.

Reviewer: indent by one more space to be a 4 space indent for a sub-bullet

1. Check all the schema files for correct structure

1. Execute all specified tests using testdriver with the executors.

1. Evaluate all test output files against schema for test results.

1. Run the verifier on all testOutputs, creating testReports

### Running individual executors and debugging

Running  all  of  the  steps  above  may  not  be  needed  for  development  and

> debugging. To facilitate quicker coding and testing, it may be useful to:

Reviewer: ```suggestion
debugging. To facilitate quicker coding and testing, it may be useful to run using the 100 test case sample set.
```

The file run_config.json is simply a list of configuration information for
running selected versions of a platform with selected test types
(components). Here's an entry for ICU4C in version 76 that runs 7 components:

> ````

Reviewer: 3 not 4 backticks

optional: change "```" to "```json" to get JSON syntax highlighting

[
  {
    "prereq": {
      "name": "Get ICU4C 76",
      "version": "76.1",
      "command": "bash ../executors/cpp/set_icu4c_binary.sh ../gh-cache/icu4c-76_1-Ubuntu22.04-x64.tgz"
    },
    "run": {
      "icu_version": "icu76",
      "exec": "cpp",
      "test_type": [
        "collation_short",
        "datetime_fmt",
        "lang_names",
        "likely_subtags",
        "message_fmt2",
        "number_fmt",
        "plural_rules"
      ],
      "per_execution": 10000
    }
  },
  ...
]

> ````

Reviewer: 3 not 4 backticks

The section *prereq* is run before this particular executor `cpp` is run under
`testdriver`. The value of `per_execution` deteremines how many tests are passed
to a single instantiation of a test executor.

To use quicker development mode, do the following:

1. Clone run_config.json. Include only the executors and/or components that are
   being developed. Also consider selecting only the ICU version or versions
   needed.

1. Close one of the executions scripts to create a custom version. Change the
   reference from `run_config.json` to the modified version. Note
   that `export TEST_LIMIT=` may be changed to test a smaller set of cases. Note
   that this may also point to a custom directory for the output under the
   conformance directory.

To use this, simply execute the custom script using `bash` or equivalent in the
environment:

````
bash my_custom_script.sh

> ````

Reviewer: 3 not 4 backticks

  results from the execution of each test in a given platform. This data may
  include output from the executor including platform-specific parameters and
  settings derived from the input data. It may also include data on errors
  encountered during the run of the platform executor.

* **Schema files** describing the expected JSON format of each type of files for
  the components.

Schema validation is performed at these points in standard processing:

1. After test data generation, all generated test data and expected result data
   files are checked for correct structure.

2. Before test execution, the schema files themselves are checked for correct
   schema structure.

3. After test executors are run, all resulting test output files are checked
   for correct structure.

### Schema file directory structure
The top level directory `schema` contains the following:

* Python routines for checking these types of data.

* A Python routine for validating the structure of the .json schema files.

* One subdirectory for each component such as "collation". This contains schema
  .json files for generated tests, expected results, and test output structure.

> ```

Reviewer: optional: change to "```bash"

crash.

# How to update Conformance Testing: ICU versions, platforms, components

Data Driven Testing is expected to remain current with ICU programs and data
updates. It is also designed to support new testing platforms in addition to the
current set of Dart, ICU4C, ICU4J, ICU4X, and NodeJS. And new types of tests,
i.e., "components", may be added to Conformance testing.

This section describes the process for keeping DDT up to date with needed test
types and required programming platforms

## Incorporating new ICU  / CLDR versions into DDT

ICU releases are usually made twice each calendar year, incorporating new data,
fixes, and new test files. ICU versions may also add new types of data
processing. A recent example is Message Format 2.

Because Data Driven Testing operations with multiple ICU and CLDR versions, this
system should be updated with each new ICU release. Here are several pull
requests for recent ICU updates:

* [ICU 76 for C++](https://github.com/unicode-org/conformance/pull/325/)

* [ICU 76 for Java](https://github.com/unicode-org/conformance/pull/344)

* [ICU76 for NodeJS](https://github.com/unicode-org/conformance/pull/348)

### ICU4C updates

> These are usually the first changes to be made because ICU4C includes both code
> and test data updates for many components.

Reviewer: ```suggestion
These are usually the first changes to be made because ICU4C includes updates to code,
locale data, and test data for many components.
```

1. Test Driver:
* Add new ICU version data in several places in testdriver/datasets.py

2. testgen:
* Add a new directory for the icu version under testgen, e.g., icu76

> * In this directory, copy test data from sources including icu4c/source. Thes

Reviewer: ```suggestion
* In this directory, copy test data from sources including icu4c/source. These
```

>   files includ collation tests, number format data, and others.

Reviewer: ```suggestion
  files include collation tests, number format data, and others.
```

      "command": "nvm install 23.3.0;nvm use 23.3.0 --silent"
    },
    "run": {
      "icu_version": "icu76",
      "exec": "node",
      "test_type": [
        "collation_short",
        "datetime_fmt",
        "list_fmt",
        "number_fmt",
        "lang_names",
        "likely_subtags",
        "rdt_fmt",
        "plural_rules"
      ],
      "per_execution": 10000
    }
  },
````

### Update ICU4J /Java to new ICU version

** TBD **
This requires referencing the new ICU4J versions in Maven Central (!!! REFERENCE NEEDED !!!)

#### run_config.json additions for Java

Updates to this file are straightforward.

### Update ICU4X / Rust to new ICU version

** TBD **

ICU4X is actively updating APIs in each new version. ICU4X releases are not
closely coordinated with ICU versions.

Adding a new ICU4X version after 1.4 may require significant changes to existing
#### run_config.json additions for ICU4X

Updates to this file are straightforward.

### Update Dart with new ICU versions

** TBD **

#### Test generator updates
Expected values for tests are obtained from several places:

* the preferred sources are generated directly from CLDR data, not mediated by

>   ICU libraries. These test files are generated by programs run wining the CLDR

Reviewer: ```suggestion
  ICU libraries. These test files are generated by programs run within the CLDR
```

>   directories and are updated with each CLDR reelease.

Reviewer: ```suggestion
  directories and are updated with each CLDR release.
```

* The second is test data used in ICU4C or ICU4J testing. For example, ICU4C
  includes 3 test files for collation that are then used by all platforms.

* Two types of test data are currently generated by NodeJS functions:
  * list format
  * relative date time format

Because of this, ICU version updated tests for these two components cannot be
run before adding a version of NodeJS that includes the new ICU version.

When the new NodeJS is incorporated into DDT, add the new NodeJS reference to
the list `icu_nvm_versions` in these files:

* testgen/generators/list_fmt.py
* testgen/generators/relativedatetime_fmt.py

## Adding New Test Types / Components
ICU supports a wide range of formatting and other functions. Many are candidates

> for Cornformance Testing. Although each has specific needs for testing, this

Reviewer: ```suggestion
for Conformance Testing. Although each has specific needs for testing, this
```

   testgen/generators. Resulting .json files with test and verification data
   should be installed in testgen/icuXX directories as needed.

* Create python modules in testgen/generators/ to read raw test data, then
  create .json file with tests and expected resuls.

* Update testgen/tesdata_gen.py with:
* Import new test generator modules
* Add new Enum values
* Add code to execute* the new generator modules

2. Define new test types in these testdriver files:
* datasets.py
* ddtargs.py
* testdriver.py
* testplan.py

3. Executors: For each executor to run the new tests:
* Add a new code file to run the tests in the executor directory, e.g.,
  `executors/cpp`

* Update configuration information such as makefiles to include the new testing
  code

* Include calling the new test routines in the main program, e.g,. `main.cpp`

Hint: Run the executor as a standalone test version, giving sample tests on the
command line or in structured test code (i.e., ICU4J's framework.)

Once the executor is working with the new test type, incorporated it into
the full execution pipeline.

4. Update run_config.json with the new test_type in each executor supporting
   the component.

For reference,

> [PR#183](https://github.com/unicode-org/conformance/pull/183/files)added these

Reviewer: ```suggestion
[PR#183](https://github.com/unicode-org/conformance/pull/183/files) added these
```

components for datetime, list format, and relative date time format.

## Adding new test platforms, e.g., types of libraries

As additional test platforms and libraries support all or part of the ICU / CLDR
functions, including them in Conformance Testing will show the degree of
compatibility of actual execution.

See [Add Dart to executors
PR#65](https://github.com/unicode-org/conformance/pull/65) for an example.

See also the
[Rust executor for ICU4x 1.3 in PR#108](https://github.com/unicode-org/conformance/pull/108)

Adding a new platform involves several changes to the DDT system:
* Change the workflow to reference the new platform

* Create a new directory structure under executors/. Add .gitignore as needed.

* Add configuration specific to the platform in the new directory under executors/

* Set up a main program that will receive instructions on the STDIN command line

    * Parse the incoming JSON data to determine test type

    * Build separate files for running each type of test

    * Return results from each testing routine in JSON format

* Support the commands for information:

>   * ** #VERSION**
>   * **#TEST**
>   *** etc.

Reviewer: ```suggestion
    * `#VERSION`
    * `#TEST`
    * etc.
```

* Update `testdriver/datasets.py` to include the new executor platform.

Note: it is very helpful to include sets of tests for the new platform for each
supported component. The ICU4J model with Intellij is a good example.

Make sure that your new executor can be run from a debugging environment or from
the command line. Tests can be tested easily here before adding them to the full execution scripts.

* Add the new platform to `run_config.json` including required setup and the implemented
  components for the DDT workflow.

See [Add Dart to executors
PR#65](https://github.com/unicode-org/conformance/pull/65) for am example.

See also the [Rust executor for ICU4x 1.3 in
PR#108](https://github.com/unicode-org/conformance/pull/108)

## Conformance Data models

> ### Directory **testData**:

Reviewer: ```suggestion
### Directory `testData`:
```

Conformance Testing uses JSON formatted data files to describe tests and
parameters as well as to record test results and deliver information for the results dashboard. 

Test generation creates the test and verify data files for each version of ICU
in .json format:

  * A test data file for each type of test. Each contains a list of labeled
  tests with parameters, options, and input values for computing output
  strings.

  * Verify files for each test type. Each contains expected results for each
    test case.

> For example, here is the structore for directory **toplevel**:

Reviewer: ```suggestion
For example, here is the structure for directory `toplevel`:
```

```
toplevel/testData/
├── icu67
│   └── ...
├── icu68
│   └── ...
...
├── icu76
│   ├── collation_test.json
│   ├── collation_verify.json
│   ├── datetime_fmt_test.json
│   ├── datetime_fmt_verify.json
│   ├── lang_name_test_file.json
│   ├── lang_name_verify_file.json
│   ├── likely_subtags_test.json
│   ├── likely_subtags_verify.json
│   ├── list_fmt_test.json
│   ├── list_fmt_verify.json
│   ├── message_fmt2_test.json
│   ├── message_fmt2_verify.json
│   ├── numberformattestspecification.txt
│   ├── numberpermutationtest.txt
│   ├── num_fmt_test_file.json
│   ├── num_fmt_verify_file.json
│   ├── plural_rules_test.json
│   ├── plural_rules_verify.json
│   ├── rdt_fmt_test.json
│   └── rdt_fmt_verify.json

```

> ### Directory **testOutput**

Reviewer: ```suggestion
### Directory `testOutput`
```
